File: src/main/java/io/dddrive/enums/adapter/api/jsonapi/dto/EnumeratedDto.java
package io.dddrive.enums.adapter.api.jsonapi.dto;

import io.crnk.core.resource.annotations.JsonApiId;
import io.dddrive.ddd.model.Aggregate;
import io.dddrive.doc.model.Doc;
import io.dddrive.enums.model.Enumerated;
import io.dddrive.obj.model.Obj;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data()
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class EnumeratedDto {

	@JsonApiId
	private String id;

	private String name;

	private EnumeratedDto itemType;

	public static EnumeratedDto fromEnum(Enumerated e) {
		if (e == null) {
			return null;
		}
		return EnumeratedDto.builder()
				.id(e.getId())
				.name(e.getName())
				.build();
	}

	public static EnumeratedDto fromAggregate(Aggregate a) {
		if (a == null) {
			return null;
		}
		return EnumeratedDto.builder()
				.id(a.getId().toString())
				.itemType(fromEnum(a.getMeta().getAggregateType()))
				.name(a.getCaption())
				.build();
	}

	public static EnumeratedDto fromObj(Obj obj) {
		return fromAggregate(obj);
	}

	public static EnumeratedDto fromDoc(Doc doc) {
		return fromAggregate(doc);
	}

}

File: fm-domain/src/main/java/io/zeitwert/dddrive/ddd/adapter/api/jsonapi/dto/EnumeratedDto.kt
package io.zeitwert.dddrive.ddd.adapter.api.jsonapi.dto

import dddrive.app.ddd.model.Aggregate
import dddrive.ddd.core.model.Part
import dddrive.ddd.enums.model.Enumerated

data class EnumeratedDto(
	val id: String,
	val name: String? = null,
) {

	override fun toString(): String = "[$id: ${name ?: "__"}]"

	companion object {

		@JvmStatic
		fun of(
			id: String,
			name: String,
		): EnumeratedDto = EnumeratedDto(id, name)

		@JvmStatic
		fun of(e: Enumerated?): EnumeratedDto? = e?.let { EnumeratedDto(it.id, it.defaultName) }

		@JvmStatic
		fun of(a: Aggregate?): EnumeratedDto? = a?.let { EnumeratedDto(DtoUtils.idToString(it.id)!!, it.caption) }

		@JvmStatic
		fun of(
			p: Part<*>?,
			name: String?,
		): EnumeratedDto? = p?.let { EnumeratedDto(it.id.toString(), name) }
	}

}
